Improve setting a global keyboard shortcut

karaggeorge opened the pull request:
Closes #868 

So, this rewrites our handling of accelerators a bit. First, it allows for a bit of better handling of the input (where some keys wouldn't work or show weird characters).

Second, it supports pasting a custom accelerator for edge cases like `Command+Shift+5` where even though we are able to register it, we can't receive it through our input (since the system won't trigger the `mouseDown` event)

Finally, it adds support for `accelerator` custom type for plugins config.

For #868, the only workaround I was able to think of (other than manually editing the `config.json` file), is that you can copy the `Command+Shift+5` text, and then paste in the input.

sindresorhus commented:
Can you describe the PR?

For example, in https://github.com/wulkano/Kap/issues/868#issuecomment-649712117 you said it was not possible. What changed?

karaggeorge commented:
@sindresorhus Yeah, I was still working some stuff out. Updated the description

karaggeorge commented on renderer/components/preferences/shortcut-input.js:
```diff
@@ -35,176 +36,201 @@ Key.propTypes = {
+  ['Alt', '⌥'],
+  ['Option', '⌥'],
+  ['Shift', '⇧'],
+  ['Cmd', '⌘'],
```
Well, the reason I decided to support all was that I allowed for pasting as described in the description. The validator will automatically change it to use Command but when you first paste it, we still want it to show the right keys.

sindresorhus replied:
You should not use `which`, use the modern alternative https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/code or https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key

karaggeorge replied:
Yeah that was part of the reason I re-wrote it. We were using `which` for everything. I used `key`. `code` is not defined. I think there was `keyCode` and `charCode` available. The only time I used the `which` field was for when `Option` is used where the `key` field has some weird characters

karaggeorge replied:
I went with `keyCode` instead

sindresorhus replied:
What do you mean `code` is not defined? It's been available since Chrome 48.

karaggeorge replied:
When I console.log `event.code` on the `onKeyDown` handler, I get `undefined` 🤔

sindresorhus replied:
React doesn't provide the real event, but it looks like it should be supported: https://github.com/facebook/react/pull/18287

karaggeorge replied:
Looks like it was merged on April 4th, last React release was on March 19th, so it's not released yet

sindresorhus replied:
Can you add a TODO comment about changing it when the next React version is out?

karaggeorge commented on main/utils/accelerator-validator.js:
```diff
@@ -0,0 +1,129 @@
+'use strict';
+
```
Well there are some specific rules about what accelerators are valid and I wanted them to be handled. Also the `event.which` and `event.code` work really weirdly with some meta keys (like option or shift). Also, with the current implementation you weren't able to register num keys, as we used the `event.which` which would register for example `Cmd+5` instead of `Cmd+num5`

For all these edge cases, I decided to make a full validation utility based on electron's docs

sindresorhus replied:
I would write a couple of short sentences about it as a code comment on the top of this file. Just for future readers.

karaggeorge commented on main/common/settings.js:
```diff
@@ -6,6 +6,16 @@ const Store = require('electron-store');
+const shortcutToAccelerator = require('../utils/shortcut-to-accelerator');
+
+const shortcuts = {
+  triggerCropper: 'Trigger Kap'
```
It doesn't close the cropper. It triggers the cropper if it's closed and then if you are recording, it stops the recording. I'm okay with renaming, I just used the title we had in the preferences until now

sindresorhus commented on docs/plugins.md:
```diff
@@ -347,6 +347,25 @@ const config = {
+const config = {
+  shortcut: {
+    title: 'Trigger Shortcut',
+    customType: 'accelerator',
```
Let's call it `keyboardShortcut`. Accelerator is not a common term for this.

karaggeorge replied:
Renamed 👍

sindresorhus commented on renderer/components/config/tab.js:
```diff
@@ -7,13 +7,31 @@ import Select from '../preferences/item/select';
+      <div>
+        <ShortcutInput
+          shortcut={value} onChange={value => {
+            console.log(name, value);
```
Forgot to remove?

sindresorhus commented on renderer/components/config/tab.js:
```diff
@@ -7,13 +7,31 @@ import Select from '../preferences/item/select';
+          }}/>
+        <style jsx>{`
+          div {
+
```
?

sindresorhus commented:
> Well, the reason I decided to support all was that I allowed for pasting as described in the description.

I don't think it's immediately obvious to the user that they can paste in the shortcut if they cannot type it in, nor what format it should be in. We need to somehow inform the user about this.

sindresorhus commented on renderer/components/preferences/shortcut-input.js:
```diff
@@ -35,176 +36,203 @@ Key.propTypes = {
+    resetKeys();
+  }, [shortcut]);
+
+  const keysToRender = keys.map(key => {
+    return metaCharacters.get(key) || key;
+  });
```
```suggestion
  const keysToRender = keys.map(key => metaCharacters.get(key) || key);
```

sindresorhus commented on renderer/containers/preferences.js:
```diff
@@ -244,7 +245,14 @@ export default class PreferencesContainer extends Container {
     try {
       await ipc.callMain('update-shortcut', {setting, shortcut});
-      this.setState({[setting]: shortcut});
+      this.setState({
```
Just use the callback version of `setState`.

karaggeorge replied:
We are on Unstated v1, which doesn't support the callback style. While doing some work on refactoring the cropper, editor and exports, I'm trying to see if we can start a general refactoring, including switching to unstated-next and hooks whenever possible, but that still needs work

sindresorhus replied:
👍

sindresorhus commented:
I get:

> general.js:119 Uncaught TypeError: Cannot read property 'triggerCropper' of undefined

When trying to open Preferences.

karaggeorge commented:
> I don't think it's immediately obvious to the user that they can paste in the shortcut if they cannot type it in, nor what format it should be in. We need to somehow inform the user about this.

Yeah I realized that too, but I'm not sure how to note that. It's really only for like `Cmd+Shift+[1-7]`, so I figured if it was documented in that issue, anyone who runs into that exact problem will see the "workaround".

The preferences error is fixed 😄

sindresorhus commented:
> Yeah I realized that too, but I'm not sure how to note that. It's really only for like Cmd+Shift+[1-7], so I figured if it was documented in that issue, anyone who runs into that exact problem will see the "workaround".

I think it will affect many, as it's common to want to use the normal screenshot keyboard shortcuts for this. Maybe we could add a small `(?)` type button on the edge of the input that when clicked tell the user about the pasting trick?

karaggeorge commented:
Could it be just a small span with `?`, maybe in a circle and that has a tooltip? We don't have a popup style component

sindresorhus commented:
Sure

karaggeorge commented:
What should the copy be? Something to explain that you can paste a valid Electron accelerator string. Maybe offer the `Command+Shift+5` as an example.

Thinking a different solution might be to have a little arrow there, clicking it would open a Menu with these most common shortcuts that you can't enter (Command+Shift+[1-7]), selecting one would add it as the shortcut?

sindresorhus commented:
I would do both. We cannot predefine all untypable keyboard shortcuts, but we can do so with the common ones.

---

> Something to explain that you can paste a valid Electron accelerator string. Maybe offer the Command+Shift+5 as an example.

👍🏻

karaggeorge commented:
Updated. Discussed with @skllcrn and added the tooltip to the top-level keyboards shortcut item, so it won't be repeated if we add more later on.

Also added a dropdown with the ones I can't add. I tested again it looks like the only ones you can't enter are `Cmd+Shift[3-6]`, so I added those

![Screen Shot 2020-07-30 at 12 23 01 PM](https://user-images.githubusercontent.com/8822835/88948519-0624f480-d260-11ea-96f8-831fb31b91fb.png)
![Screen Shot 2020-07-30 at 12 22 58 PM](https://user-images.githubusercontent.com/8822835/88948522-06bd8b00-d260-11ea-8357-f32f44641b65.png)

sindresorhus commented:
Nice work!

sindresorhus commented:
I noticed one thing that could be improved, but I don't think it's a new issue. If there's already a shortcut, you focus the input and then press Esc, it should just blur the input, not clear the shortcut. Another thing, if I press Shift, and then release Shift, the input field shakes, but I don't think it should when the shortcut is incomplete.
